icu4c/source/tools/toolutil/writesrc.cpp

```diff
@@ -357,6 +358,69 @@ usrc_writeUCPTrie(FILE *f, const char *name, const UCPTrie *pTrie, UTargetSyntax
     usrc_writeUCPTrieStruct(f, line, pTrie, line2, line3, line4, syntax);
 }
 
+U_CAPI void U_EXPORT2
+usrc_writeUnicodeSet(
+        FILE *f,
+        const char *name,
+        const USet *pSet,
+        UTargetSyntax syntax) {
+    // ccode is not yet supported
+    U_ASSERT(syntax == UPRV_TARGET_SYNTAX_TOML);
+
+    fprintf(f, "name = \"%s\"\n", name);
+
+    // First write as the uset serialized format
+    icu::ErrorCode localStatus;
+    int32_t length = uset_serialize(pSet, NULL, 0, localStatus);
+    localStatus.reset();
+    std::unique_ptr<uint16_t[]> arr(new uint16_t[length]);
+    uset_serialize(pSet, arr.get(), length, localStatus);
+    U_ASSERT(localStatus.isSuccess());
+    const char* indent = (syntax == UPRV_TARGET_SYNTAX_TOML) ? "  " : "";
+    usrc_writeArray(f, "serialized = [\n  ", arr.get(), 16, length, indent, "\n]\n");
+
+    // Now write it also as a list of ranges
+    USerializedSet serializedSet;
+    uset_getSerializedSet(&serializedSet, arr.get(), length);
+    fprintf(f, "ranges = [\n");
+    for (int32_t i=0; i<uset_getSerializedRangeCount(&serializedSet); i++) {
+        UChar32 start;
+        UChar32 end;
+        uset_getSerializedRange(&serializedSet, i, &start, &end);
+        fprintf(f, "  [0x%x, 0x%x],\n", start, end);
+    }
+    fprintf(f, "]\n");
+}
+
+U_CAPI void U_EXPORT2
+usrc_writeUCPMap(
+        FILE *f,
+        const char *name,
+        const UCPMap *pMap,
+        UProperty uproperty,
+        UTargetSyntax syntax) {
+    // ccode is not yet supported
+    U_ASSERT(syntax == UPRV_TARGET_SYNTAX_TOML);
+
+    fprintf(f, "name = \"%s\"\n", name);
+
+    // Print out list of ranges
+    UChar32 start = 0, end;
+    uint32_t value;
+    fprintf(f, "ranges = [\n");
+    while ((end = ucpmap_getRange(pMap, start, UCPMAP_RANGE_NORMAL, 0, NULL, NULL, &value)) >= 0) {
+        if (uproperty != UCHAR_INVALID_CODE) {
+            const char* name1 = u_getPropertyValueName(uproperty, value, U_LONG_PROPERTY_NAME);
+            const char* name2 = u_getPropertyValueName(uproperty, value, U_SHORT_PROPERTY_NAME);
+            fprintf(f, "  [0x%x, 0x%x, %d, \"%s\", \"%s\"],\n", start, end, value, name1, name2);
+        } else {
+            fprintf(f, "  [0x%x, 0x%x, %d],\n", start, end, value);
+        }
```

Comment:
Optional: By the way, since you print all ranges, the end of one and the start of the next always differ by exactly one.
This is fine for human readability, but redundant for data. (An inversion map would only have the *limit* of each range and the range's value -- good for binary search but not nice for humans. Or print only the range start but not limit nor end.)

Reply:
In these TOML files I would rather optimize for human readability, because these are pre-processed again in ICU4X.